kasa/iot/modules/motion.py

```diff
@@ -48,44 +116,246 @@ def _initialize_features(self) -> None:
         return self.data["get_config"]
 
     @property
-    def range(self) -> Range:
-        """Return motion detection range."""
-        return Range(self.config["trigger_index"])
+    def pir_config(self) -> PIRConfig:
+        """Return PIR sensor configuration."""
+        pir_range = Range(self.config["trigger_index"])
+        return PIRConfig(
+            enabled=bool(self.config["enable"]),
+            adc_min=int(self.config["min_adc"]),
+            adc_max=int(self.config["max_adc"]),
+            range=pir_range,
+            threshold=self.get_range_threshold(pir_range),
+        )
 
     @property
     def enabled(self) -> bool:
         """Return True if module is enabled."""
-        return bool(self.config["enable"])
+        return self.pir_config.enabled
+
+    @property
+    def adc_min(self) -> int:
+        """Return minimum ADC sensor value."""
+        return self.pir_config.adc_min
+
+    @property
+    def adc_max(self) -> int:
+        """Return maximum ADC sensor value."""
+        return self.pir_config.adc_max
+
+    @property
+    def adc_mid(self) -> int:
+        """
+        Return the midpoint for the ADC.
+
+        The midpoint represents the zero point for the PIR sensor waveform.
+
+        Currently this is estimated by:
+            math.floor(abs(adc_max - adc_min) / 2)
+        """
+        return self.pir_config.adc_mid
 
     async def set_enabled(self, state: bool) -> dict:
         """Enable/disable PIR."""
         return await self.call("set_enable", {"enable": int(state)})
 
-    async def set_range(
-        self, *, range: Range | None = None, custom_range: int | None = None
-    ) -> dict:
-        """Set the range for the sensor.
+    @property
+    def ranges(self) -> list[str]:
+        """Return set of supported range classes."""
+        range_min = 0
+        range_max = len(self.config["array"])
+        valid_ranges = list()
+        for r in Range:
+            if (r.value >= range_min) and (r.value < range_max):
+                valid_ranges.append(r.name)
+        return valid_ranges
+
+    @property
+    def range(self) -> Range:
+        """Return motion detection Range."""
+        return self.pir_config.range
+
+    async def set_range(self, range: Range) -> dict:
+        """Set the Range for the sensor.
 
-        :param range: for using standard ranges
-        :param custom_range: range in decimeters, overrides the range parameter
+        :param Range: the range class to use.
         """
-        if custom_range is not None:
-            payload = {"index": Range.Custom.value, "value": custom_range}
-        elif range is not None:
-            payload = {"index": range.value}
-        else:
-            raise KasaException("Either range or custom_range need to be defined")
+        payload = {"index": range.value}
+        return await self.call("set_trigger_sens", payload)
+
+    def _parse_range_value(self, value: str) -> Range:
```

Comment:
Now that the library is python >= 3.11 could this all be made simpler by converting `Range` to `StrEnum`?

Reply:
From what I can tell, `StrEnum` will not improve this use case unfortunately, as it still does not define a proper `Enum.get(key, default)`.

The enum is also a mapping of the integer returned by `trigger_index`, so perhaps we could use `IntEnum`? But then it comes down to how you want the range to be presented to the user, as `0, 1, 2, 3` or `FAR, NEAR, MID, CUSTOM`. Currently it will accept the range enum names, in upper or lower case, but of course that relies on the range classes not changing on future devices.

Reply:
I don't completely follow the `Enum.get(key, default)`. Can't you just catch a value error? StrEnum would mean the method could take `str | StrEnum` which could mean more flexibility when we have a common interface across devices.